pkg/platform/kube/ingress/ingress.go

```diff
@@ -0,0 +1,423 @@
+	}
+
+	if spec.SSLPassthrough {
+		ingressAnnotations["nginx.ingress.kubernetes.io/ssl-passthrough"] = "true"
+	} else {
+		var authIngressAnnotations map[string]string
+
+		switch spec.AuthenticationMode {
+		case AuthenticationModeNone:
+			//do nothing
+		case AuthenticationModeBasicAuth:
+			authIngressAnnotations, secretResource, err = im.getBasicAuthIngressAnnotationsAndSecret(ctx, spec)
+			if err != nil {
+				return nil, errors.Wrap(err, "Failed to get basic auth annotations")
+			}
+		case AuthenticationModeAccessKey:
+
+			// relevant when running on iguazio platform
+			authIngressAnnotations, err = im.getSessionVerificationAnnotations("/api/data_sessions/verifications")
+			if err != nil {
+				return nil, errors.Wrap(err, "Failed to get access key auth mode annotations")
+			}
+		case AuthenticationModeDex:
+			authIngressAnnotations, err = im.getDexAuthIngressAnnotations(spec)
+			if err != nil {
+				return nil, errors.Wrap(err, "Failed to get dex auth annotations")
+			}
+		default:
+			return nil, errors.Errorf("Unknown ingress authentication mode: %s", spec.AuthenticationMode)
+		}
+
+		// merge with existing annotation map
+		for annotation, annotationValue := range authIngressAnnotations {
+			ingressAnnotations[annotation] = annotationValue
+		}
+
+		ingressAnnotations["nginx.ingress.kubernetes.io/proxy-body-size"] = "0"
+
+		// redirect to SSL if spec allows it, and given the system is configured to not allow HTTP in allowed-protocols
+		if spec.AllowSSLRedirect && len(spec.AllowedProtocols) == 1 && spec.AllowedProtocols[0] == "https" {
+			ingressAnnotations["nginx.ingress.kubernetes.io/ssl-redirect"] = "true"
+		} else {
+			ingressAnnotations["nginx.ingress.kubernetes.io/ssl-redirect"] = "false"
+		}
+	}
+
+	if spec.ProxyReadTimeout != "" {
+		ingressAnnotations["nginx.ingress.kubernetes.io/proxy-read-timeout"] = spec.ProxyReadTimeout
+	}
+
+	if spec.WhitelistIPAddresses != nil {
+		ingressAnnotations["nginx.ingress.kubernetes.io/whitelist-source-range"] = strings.Join(spec.WhitelistIPAddresses, ",")
+	}
+
+	if spec.Annotations != nil {
+		for annotation, annotationValue := range spec.Annotations {
+			ingressAnnotations[annotation] = annotationValue
+		}
+	}
+
+	ingressResource := &v1beta1.Ingress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: spec.Name,
+			Namespace: spec.Namespace,
+			Annotations: ingressAnnotations,
+		},
+		Spec: v1beta1.IngressSpec{
+			Rules: []v1beta1.IngressRule{
+				{
+					Host: spec.Host,
+					IngressRuleValue: v1beta1.IngressRuleValue{
+						HTTP: &v1beta1.HTTPIngressRuleValue{
+							Paths: []v1beta1.HTTPIngressPath{
+								{
+									Path: spec.Path,
+									Backend: v1beta1.IngressBackend{
+										ServiceName: spec.ServiceName,
+										ServicePort: intstr.FromInt(spec.ServicePort),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	// if no specific TLS secret was given - set it to be system's TLS secret
+	tlsSecret := spec.TLSSecret
```

Comment:
lines 152-166 can be moved to `populateTLSSecret()`, wdyt?

Reply:
Could shorten the code a little bit, but I think we should only manipulate the ingressResource object inside this function to keep it simple